tianshou/policy/modelfree/redq.py

```diff
@@ -0,0 +1,207 @@
+        else:
+            self._alpha = alpha
+
+        if target_mode in ('min', 'mean'):
+            self.target_mode = target_mode
+        else:
+            raise ValueError('Unsupported mode of Q target computing.')
+
+        self.critic_gradient_step = 0
+        self.actor_delay = actor_delay
+        self._deterministic_eval = deterministic_eval
+        self.__eps = np.finfo(np.float32).eps.item()
+
+    def train(self, mode: bool = True) -> "REDQPolicy":
+        self.training = mode
+        self.actor.train(mode)
+        self.critics.train(mode)
+        return self
+
+    def sync_weight(self) -> None:
+        for o, n in zip(self.critics_old.parameters(), self.critics.parameters()):
+            o.data.copy_(o.data * (1.0 - self._tau) + n.data * self._tau)
+
+    def forward(  # type: ignore
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        obs = batch[input]
+        logits, h = self.actor(obs, state=state, info=batch.info)
+        assert isinstance(logits, tuple)
+        dist = Independent(Normal(*logits), 1)
+        if self._deterministic_eval and not self.training:
+            act = logits[0]
+        else:
+            act = dist.rsample()
+        log_prob = dist.log_prob(act).unsqueeze(-1)
+        # apply correction for Tanh squashing when computing logprob from Gaussian
+        # You can check out the original SAC paper (arXiv 1801.01290): Eq 21.
+        # in appendix C to get some understanding of this equation.
+        if self.action_scaling and self.action_space is not None:
+            action_scale = to_torch_as(
+                (self.action_space.high - self.action_space.low) / 2.0, act
+            )
+        else:
+            action_scale = 1.0  # type: ignore
+        squashed_action = torch.tanh(act)
+        log_prob = log_prob - torch.log(
+            action_scale * (1 - squashed_action.pow(2)) + self.__eps
+        ).sum(-1, keepdim=True)
+        return Batch(
+            logits=logits, act=squashed_action, state=h, dist=dist, log_prob=log_prob
+        )
+
+    def _target_q(self, buffer: ReplayBuffer, indices: np.ndarray) -> torch.Tensor:
+        batch = buffer[indices]  # batch.obs: s_{t+n}
+        obs_next_result = self(batch, input='obs_next')
+        a_ = obs_next_result.act
+        sample_ensemble_idx = np.random.choice(
+            self.ensemble_size, self.subset_size, replace=False
+        )
+        qs = self.critics_old(batch.obs_next, a_)[sample_ensemble_idx, ...]
+        if self.target_mode == 'min':
+            target_q, _ = torch.min(qs, dim=0)
+        elif self.target_mode == 'mean':
+            target_q = torch.mean(qs, dim=0)
+        target_q -= self._alpha * obs_next_result.log_prob
+
+        return target_q
+
+    def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
+        # critic ensemble
```

Comment:
Is it possible to call `super().learn()` here to reduce code duplication? If so, is it possible to make `REDQPolicy` a decorator so that it can be used with any off-policy algorithm?

Reply:
It is possible if the training of actor and critic are written in helper methods. Then we can overwrite train_critic method in REDQ. As for the decorator idea, I don't have a clear thought of solution at this moment.